localboot/grub.go

```diff
@@ -95,35 +124,53 @@ func ParseGrubCfg(grubcfg string, basedir string, grubVersion int) []bootconfig.
 	return bootconfigs
 }
 
+func unquote(ver grubVersion, text string) string {
+	if ver == grubV2 {
+		// if grub2, unquote the string, as directives could be quoted
+		// https://www.gnu.org/software/grub/manual/grub/grub.html#Quoting
+		// TODO unquote everything, not just \$
+		return strings.Replace(text, `\$`, "$", -1)
+	}
+	// otherwise return the unmodified string
+	return text
+}
+
 // ScanGrubConfigs looks for grub2 and grub legacy config files in the known
 // locations and returns a list of boot configurations.
 func ScanGrubConfigs(basedir string) []bootconfig.BootConfig {
 	bootconfigs := make([]bootconfig.BootConfig, 0)
-	// Scan Grub 2 configurations
-	for _, grubpath := range Grub2Paths {
-		path := path.Join(basedir, grubpath)
-		log.Printf("Trying to read %s", path)
-		grubcfg, err := ioutil.ReadFile(path)
-		if err != nil {
-			log.Printf("cannot open %s: %v", path, err)
-			continue
+	err := filepath.Walk(basedir, func(currentPath string, info os.FileInfo, err error) error {
+		if path.Dir(currentPath) == basedir && info.IsDir() && !isGrubSearchDir(path.Base(currentPath)) {
+			log.Printf("Skip %s", currentPath)
+			return filepath.SkipDir // skip irrelevant toplevel directories
 		}
-		crypto.TryMeasureData(crypto.ConfigData, grubcfg, path)
-		cfgs := ParseGrubCfg(string(grubcfg), basedir, 2)
-		bootconfigs = append(bootconfigs, cfgs...)
-	}
-	// Scan Grub Legacy configurations
-	for _, grubpath := range GrubLegacyPaths {
-		path := path.Join(basedir, grubpath)
-		log.Printf("Trying to read %s", path)
-		grubcfg, err := ioutil.ReadFile(path)
-		if err != nil {
-			log.Printf("cannot open %s: %v", path, err)
-			continue
+		if info.IsDir() {
+			log.Printf("Check %s", currentPath)
+			return nil // continue
+		}
+		cfgname := info.Name()
+		if cfgname == "grub.cfg" || cfgname == "grub2.cfg" {
+			var ver grubVersion
+			if cfgname == "grub.cfg" {
+				ver = grubV1
+			} else if cfgname == "grub2.cfg" {
+				ver = grubV2
+			}
+			// try parsing
+			log.Printf("Trying to read %s", currentPath)
+			grubcfg, errRead := ioutil.ReadFile(currentPath)
+			if errRead != nil {
+				log.Printf("cannot open %s: %v", currentPath, errRead)
+				return nil // continue anyway
+			}
+			crypto.TryMeasureData(crypto.ConfigDataPCR, grubcfg, currentPath)
```

Comment:
measuring boot configurations should not be part of parsing them

Reply:
done